Fix rhessi.get_obssumm_file()

m-charlton opened the pull request:
Minor problem causing get_obssumm_file not to fetch HESSI file

#### Changes

Made minor changes for the pep8 police. There are three lines in docstrings
which are longer than 100 hundred characters, but these contain very long 
URLs and I don't how to reformat these. 

##### `get_obssumm_file(...)`

Now works as expected. I've added a warning to inform the user if an attempt
is made to retrieve more than one whole day of data. This may be overkill as
this limitation is clearly stated in the docstring.

Also noticed that this check would have to be done in `get_obssumm_dbase_file(...)`,
and `get_obssum_filename(...)`. If that were the case these functions are called
be each other then you'd see the warning message more than once. I think
I've convinced my self to remove it.


##### Noticed the following

If your network connection is down, `get_base_url(...)` will return `None`

```python
def get_base_url():
    """
    Find the first mirror which is online
    """

    for server in data_servers:
        try:
            urlopen(server, timeout=1)
        except (URLError, socket.timeout):
            pass
        else:
            return server
```

Would it be safer to do the following:

```python
def get_base_url():
    """
    Find the first mirror which is online
    """

    for server in data_servers:
        try:
            urlopen(server, timeout=1)
            return server
        except (URLError, socket.timeout):
            pass

    raise IOError('Unable to find an online HESSI server from {0}'.format(data_servers))
```

Wording of exception message may need improving.

I've some ideas for unit tests. Is it worth submitting these? Or is 
the code in too fluid a state?

pep8speaks commented:
Hello @m-charlton! Thanks for submitting the PR.

 - In the file [`sunpy/instr/rhessi.py`](https://github.com/m-charlton/sunpy/blob/4f1e98f654f8da1b979228aab76e1c043c1a5706/sunpy/instr/rhessi.py), following are the PEP8 issues :

> [Line 85:101](https://github.com/m-charlton/sunpy/blob/4f1e98f654f8da1b979228aab76e1c043c1a5706/sunpy/instr/rhessi.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 132:101](https://github.com/m-charlton/sunpy/blob/4f1e98f654f8da1b979228aab76e1c043c1a5706/sunpy/instr/rhessi.py#L132): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 354:101](https://github.com/m-charlton/sunpy/blob/4f1e98f654f8da1b979228aab76e1c043c1a5706/sunpy/instr/rhessi.py#L354): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 100 characters)

Cadair commented:
> Made minor changes for the pep8 police.

The pep8speaks bot thanks you :D

I would submit the unit tests as well, if you are willing, this code is a right pain due to lack of tests.

I think raising an error if a server can't be found makes sense as well, certainly more sense than a "can't do that with None type" error.

m-charlton commented:
What do you think of the warning message if the number of days is greater than one?

Cadair commented on sunpy/instr/rhessi.py:
```diff
@@ -236,16 +238,17 @@ def get_obssumm_file(time_range):
+    num_of_days = math.ceil(time_range.days.value)
 
-    url_root = get_base_url() + data_location
+    if num_of_days > 1:
```
you should be able to do this comparison using units:
```python
if time_range.days > 1* u.day:
    ....
```

m-charlton replied:
Great, saves all that faffing about rounding it up

Cadair commented:
the warning looks good.

m-charlton commented:
In that case I'll factor it out and add it to the other functions. This does mean that in some cases
you'll see the warning message more than once.

Cadair commented:
Try and minimise them as much as possible. But not the end of the world.

m-charlton commented:
### Changes

- Added test to issue a `UserWarning` message if the time interval is greater
  than one day. The `stacklevel` set so that the message has the line number
  of where the check failed.

- Using `posixpath.join(...)` to assemble urls. It's much safer than doing it
  by 'hand' and platform agnostic.

- Made first stab at automatically building energy bands in `parse_obssumm_file(...)`,
  using ` _build_energy_bands(...)`. Please pay particular attention to the REGEX
  used to match the units. This might be dangerous to put this code here as I've
  no unit tests for `parse_obssumm_file(...)`.

- Added `sunpy/instr/tests/test_rhessi.py`. There are three 'line too long ...' warnings.
  I can't get round these as the lines in question represent the lines found in a
  RHESSI observation summary file.

#### `parse_obssumm_file(...)`

Will submit unit tests when I've figured out how to mock FITS files.

#### `parse_obssumm_hdulist(...)`

Not touched this. I'd be grateful for some examples of it's use. The return type
on the docstring is wrong. There seems a lot of duplication with `parse_obssumm_file(...)`.
Looking into refactoring.

#### `uncompress_countrate(...)`, `_backproject(...)` & `backprojection(...)`

Figuring out how these work. Will submit unit tests at a latter date.


### Observations

#### `get_obssum_filename(...)`

The API, as a whole, supports providing data from one day. This function
does not so I'm not performing the 'one day test' here. As this returns a list,
containing file name(s), should it be called `get_obssum_filenames(...)`?

Seems to be missing an 'm' in 'obssum'. Did a quick search of the code base and
found it's used in the following:

```python
# sunpy/lightcurve/sources/rhessi.py
129:        return rhessi.get_obssum_filename(time_range)[0]

# sunpy/net/dataretriever/sources/rhessi.py
24:        return rhessi.get_obssum_filename(timerange)
```
Should it be renamed? Or would this impact external code?


`lc_linecolors`, line 45, seems to be doing nothing. Could not find it used
anywhere in the code base. Though I did notice the following:

```python
# sunpy/lightcurve/sources/rhessi.py
83:        lc_linecolors = rhessi.hsi_linecolors()
85:        for lc_color, (item, frame) in zip(lc_linecolors, self.data.iteritems()):

# sunpy/timeseries/sources/rhessi.py
94:        lc_linecolors = rhessi.hsi_linecolors()
96:        for lc_color, (item, frame) in zip(lc_linecolors, self.data.iteritems()):
```

`lc_linecolors` & `rhessi.hsi_linecolors(...)` are different values but, was
`rhessi.hsi_linecolors(...)` originally meant to return `lc_linecolors`?

Cadair commented:
I am going to have to delegate to @ehsteve on most of this.

Cadair commented:
Rather than mocking fits files we normally put a small test file in the repo and use that to test loading / parsing code.

m-charlton commented:
RE: FITS files, that makes sense. I'll look at the ones that are available.

I'll wait for comments from @ehsteve.

m-charlton commented:
I've broken a test on `lightcurve/sources/rhessi.py' - investigating

m-charlton commented:
Looks like the test data is at fault

The `TestRHESSISummaryLightCurve.test_hsi_default` purpose is "Test creation with no input".
It's failing when calling `sunpy.instr.rhessi.get_obssum_filename(...)`, the
code I've been altering.

The following is the RHESSI summary file:

```
HESSI Filedb File:
Created: 2017-03-19T17:43:26.000
Number of Files:           16
                    Filename  Orb_st Orb_end         Start_time           End_time 
hsi_obssumm_20170301_044.fit   65536   65536 01-Mar-17 00:00:00 02-Mar-17 00:00:00 
hsi_obssumm_20170302_044.fit   65536   65536 02-Mar-17 00:00:00 03-Mar-17 00:00:00 
hsi_obssumm_20170303_043.fit   65536   65536 03-Mar-17 00:00:00 04-Mar-17 00:00:00 
hsi_obssumm_20170304_048.fit   65536   65536 04-Mar-17 00:00:00 05-Mar-17 00:00:00 
hsi_obssumm_20170305_050.fit   65536   65536 05-Mar-17 00:00:00 06-Mar-17 00:00:00 
hsi_obssumm_20170306_047.fit   65536   65536 06-Mar-17 00:00:00 07-Mar-17 00:00:00 
hsi_obssumm_20170307_041.fit   65536   65536 07-Mar-17 00:00:00 08-Mar-17 00:00:00 
hsi_obssumm_20170308_038.fit   65536   65536 08-Mar-17 00:00:00 09-Mar-17 00:00:00 
hsi_obssumm_20170309_043.fit   65536   65536 09-Mar-17 00:00:00 10-Mar-17 00:00:00 
hsi_obssumm_20170310_043.fit   65536   65536 10-Mar-17 00:00:00 11-Mar-17 00:00:00 
hsi_obssumm_20170311_043.fit   65536   65536 11-Mar-17 00:00:00 12-Mar-17 00:00:00 
hsi_obssumm_20170312_043.fit   65536   65536 12-Mar-17 00:00:00 13-Mar-17 00:00:00 
hsi_obssumm_20170313_043.fit   65536   65536 13-Mar-17 00:00:00 14-Mar-17 00:00:00 
hsi_obssumm_20170314_046.fit   65536   65536 14-Mar-17 00:00:00 15-Mar-17 00:00:00 
hsi_obssumm_20170315_046.fit   65536   65536 15-Mar-17 00:00:00 16-Mar-17 00:00:00 
hsi_obssumm_20170316_032.fit       0       0 16-Mar-17 00:00:00 17-Mar-17 00:00:00 
```

The tests are using recent dates, it looks as if they are dynamically generated:
```
Start: 2017-03-17 16:03:58
End:   2017-03-18 16:...8 04:03:58
```

The problem is that when we try to get the file name for 2017-03-17 , it's not there, as entries
are only available up until March 16th.

#### Suggestion

If `sunpy.instr.rhessi.get_obssum_filename(...)` cannot retrieve any URLs then 
it should raise an exception.

dpshelio commented:
Looks pretty good to me!! @ehsteve  - what do you think? you are the rhessi master in here :)

nabobalis commented:
We use mock and pytest_mock? I assume we should stick with one.

samaloney commented:
Is the plan to merge this soon? I've rebased it into branch I'm working with some additional changes.

nabobalis commented:
Maybe not this week, but maybe after 0.8 is out. We can discuss it at the dev meeting today.
